src/tools/clippy/clippy_lints/src/doc/empty_line_after.rs

```diff
@@ -0,0 +1,329 @@
+        {
+            "parent module"
+        },
+        Node::Crate(crate_mod)
+            if let Some(first) = crate_mod
+                .item_ids
+                .iter()
+                .map(|&id| cx.tcx.hir().item(id))
+                // skip prelude imports
+                .find(|item| !matches!(item.span.ctxt().outer_expn_data().kind, ExpnKind::AstPass(_)))
+                && first.owner_id == owner =>
+        {
+            "crate"
+        },
+        _ => return,
+    };
+
+    diag.multipart_suggestion_verbose(
+        match kind {
+            StopKind::Attr => format!("if the attribute should apply to the {parent_desc} use an inner attribute"),
+            StopKind::Doc(_) => format!("if the comment should document the {parent_desc} use an inner doc comment"),
+        },
+        gaps.iter()
+            .flat_map(|gap| gap.prev_chunk)
+            .map(Stop::convert_to_inner)
+            .collect(),
+        Applicability::MaybeIncorrect,
+    );
+}
+
+fn check_gaps(cx: &LateContext<'_>, gaps: &[Gap<'_>]) -> bool {
+    let Some(first_gap) = gaps.first() else {
+        return false;
+    };
+    let empty_lines = || gaps.iter().flat_map(|gap| gap.empty_lines.iter().copied());
+    let mut has_comment = false;
+    let mut has_attr = false;
+    for gap in gaps {
+        has_comment |= gap.has_comment;
+        if !has_attr {
+            has_attr = gap.prev_chunk.iter().any(|stop| stop.kind == StopKind::Attr);
+        }
+    }
+    let kind = first_gap.prev_stop.kind;
+    let (lint, kind_desc) = match kind {
+        StopKind::Attr => (EMPTY_LINE_AFTER_OUTER_ATTR, "outer attribute"),
+        StopKind::Doc(_) => (EMPTY_LINE_AFTER_DOC_COMMENTS, "doc comment"),
+    };
+    let (lines, are, them) = if empty_lines().nth(1).is_some() {
+        ("lines", "are", "them")
+    } else {
+        ("line", "is", "it")
+    };
+    span_lint_and_then(
+        cx,
+        lint,
+        first_gap.prev_stop.span.to(empty_lines().last().unwrap()),
+        format!("empty {lines} after {kind_desc}"),
+        |diag| {
+            if let Some(owner) = cx.last_node_with_lint_attrs.as_owner() {
+                let def_id = owner.to_def_id();
+                let def_descr = cx.tcx.def_descr(def_id);
+                diag.span_label(
+                    cx.tcx.def_span(def_id),
+                    match kind {
+                        StopKind::Attr => format!("the attribute applies to this {def_descr}"),
+                        StopKind::Doc(_) => format!("the comment documents this {def_descr}"),
+                    },
+                );
+            }
+
+            diag.multipart_suggestion_with_style(
+                format!("if the empty {lines} {are} unintentional remove {them}"),
+                empty_lines().map(|empty_line| (empty_line, String::new())).collect(),
+                Applicability::MaybeIncorrect,
+                SuggestionStyle::HideCodeAlways,
+            );
```

Comment:
@estebank What is the best way to suggest removing non contiguous empty lines. I.e.

```rust
1 /// Comment that should stay
2 
3 // More things that should stay
4 
5 struct X;
```

The `empty_lines()` functions returns the spans of lines `2` and `4`, but their spans are `2:1-2:1` and `4:1-4:1`, so considered empty spans. `rustfix` has no problem applying this suggestion, while the `debug_assertion` causes the failure in CI. Is there a better way to suggest removing those empty lines, than the above? If not, should the `debug_assertion` in the `multipart_suggestion` implementation be relaxed?

Reply:
If the suggestion is to remove a line, shouldn't the spans be `2:1-3:1` and `4:1-5:1`? An empty span wouldn't include the newline character at `2:1`, would it? 🤔

Reply:
I tried adding a `BytePos(1)` to the span. But then the spans overlap (for 2 contiguous empty lines), as the first span is `2:1-3:1` and the next span is `3:1-4:1` with an overlap at `3:1`. And `rustfix` can't apply multipart suggestions with overlapping spans.

Reply:
Will be fixed in https://github.com/rust-lang/rust-clippy/pull/13439

Since we already behind the sync after this one by 3 days, I will open a new PR for it.